### PR TITLE
Inclusion of qrCoder tool and its direct dependencies (qrcode and Image python libraries)

### PR DESCRIPTION
**Pull request by fabiorjvieira**
First version of qr_coder tool. 
It generates qr codes based on several parameters, however user must only fill "qr data".
See tool help section for more information or qr code library home page (https://pypi.python.org/pypi/qrcode).
Tool qr_code depends on two python libraries (qrcode and Image). Packages package_image_1_4_1 and package_qrcode_5_1 deal with these dependencies.

### CODE REVIEW

**Comment by drosofff** [comment]
I am :ok: with the pull request but wait for @mvdbeek feed back on the tool

File: packages/package_image_1_4_1/tool_dependencies.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<tool_dependency>
+            <actions>
+                <action type="setup_python_environment">
+                    <repository name="package_image_1_4_1" owner="francescortiz">
```
Comment (mvdbeek) [inline packages/package_image_1_4_1/tool_dependencies.xml]: If you want to do this the correct way you really need to use this:
```xml
         <action type="setup_python_environment">
               <repository name="package_python_2_7" owner="iuc">
                   <package name="python" version="2.7" />
               </repository>
               <package">https://codeload.github.com/francescortiz/image/tar.gz/master</package>
         </action>
```
This tells galaxy to setup the python environment use python_2_7, which will be installed for you.

File: packages/package_qrcode_5_1/tool_dependencies.xml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+            <actions>
+                <action type="download_by_url">https://pypi.python.org/packages/source/q/qrcode/qrcode-5.1.tar.gz</action>
+                <action type="make_directory">$INSTALL_DIR/lib/python</action>
+                <action type="shell_command">export PYTHONPATH=$PYTHONPATH:$INSTALL_DIR/lib/python &amp;&amp; python setup.py install --home $INSTALL_DIR --install-scripts $INSTALL_DIR/bin</action>
+                <action type="set_environment">
+                    <environment_variable name="PYTHONPATH" action="append_to">$INSTALL_DIR/lib/python</environment_variable>
+                    <environment_variable name="PYTHONPATH_QRCODE" action="set_to">$INSTALL_DIR/lib/python</environment_variable>
+                    <environment_variable name="PATH" action="prepend_to">$INSTALL_DIR/bin</environment_variable>
+                </action>
+            </actions>
```
Comment (mvdbeek) [inline packages/package_qrcode_5_1/tool_dependencies.xml]: Can you do this as you have done for package_image?
```xml
e.g.
         <action type="setup_python_environment">
               <repository name="package_python_2_7" owner="iuc">
                   <package name="python" version="2.7" />
               </repository>
               <repository name="package_image_1_4_1" owner="iuc">
                   <package name="Image" version="1.4.1" />
               </repository>
               <package">https://codeload.github.com/francescortiz/image/tar.gz/master</package>
         </action>
```

**Comment by mvdbeek** [comment]
@fabiorjvieira 
Almost ready, if you can address my comments we can merge. One last thing before that,
can you use `planemo shed_upload -t https://lbcd41.snv.jussieu.fr/toolshed/ --shed_key <your api key>` ? Then Ic an use planemo shed_test to also tests your tool dependencies ... .
Also, since you created a pip package we might as well use `<action type="setup_virtualenv">qrcode==5.1</action>`

File: packages/package_image_1_4_1/tool_dependencies.xml (hunk 1 above)
Comment (fabiorjvieira) [inline packages/package_image_1_4_1/tool_dependencies.xml]: Please, make a PR. :)

**Comment by fabiorjvieira** [comment]
1- shed_upload (in fact i did a shed_create) done.
2- two new tags added to qr_coder tool_dependencies.xml (setup_virtualenv to qrcode and image) done, but in the next commit.

File: tools/qr_coder/qr_coder.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      <param name="qrBoxSize"         type="integer" label="Controls how many pixels each 'box' of the QR code is"                     optional="true"  min="0"/>
+      <param name="qrErrorCorrection" type="select"  label="Controls the error correction used for the QR Code"                        optional="true"  >
+         <option value="Low">Low</option>
+         <option value="Medium">Medium</option>
+         <option value="Q">Q</option>
+         <option value="High">High</option>
+      </param>
+   </inputs>
+   <outputs>
+      <data name="qrOutput" label="Name of the image PNG file" format="png" />
```
Comment (mvdbeek) [inline tools/qr_coder/qr_coder.xml]: The label controls the name of the output file in the history. With this setting every file will be named "Name of the image PNG file". How about changing it to label="QR code for ${qrData}"? See https://wiki.galaxyproject.org/Admin/Tools/ToolConfigSyntax#Miscellaneous_tips_and_tricks

File: tools/qr_coder/qr_coder.xml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      #end if
+      #if $qrBoxSize
+         -qrBoxSize $qrBoxSize
+      #end if
+      #if $qrErrorCorrection
+         -qrErrorCorrection $qrErrorCorrection
+      #end if
+   </command>
+   <inputs>
+      <param name="qrData"            type="text"    label="Data to be coded as an qr image"                                           optional="false" value="https://mississippi.snv.jussieu.fr/artbio/"/>
```
Comment (mvdbeek) [inline tools/qr_coder/qr_coder.xml]: The value corresponds to what will be pre-selected. It's probably better to not have a value, which will require the user to enter something. Also note that there is a help="" tag. You could do for example help="Enter the text for which you whish to generate a QR code."

File: tools/qr_coder/qr_coder.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+   parser.add_argument("-qrVersion",         dest="qrVersion",         default="", help="An integer from 0 (automatic) to 40 that controls the size of the QR Code",     required=False)
+   #parser.add_argument("--qrFit",            dest="qrFit",             default="", help="Makes the code to determine the size of the QR Code automatically", required=False, action='store_true')
+   parser.add_argument("-qrBorder",          dest="qrBorder",          default="", help="Controls how many boxes thick the border should be",                required=False)
+   parser.add_argument("-qrBoxSize",         dest="qrBoxSize",         default="", help="Controls how many pixels each 'box' of the QR code is",             required=False)
+   parser.add_argument("-qrErrorCorrection", dest="qrErrorCorrection", default="", help="Controls the error correction used for the QR Code",                required=False, choices=[Constants.LOW, Constants.MEDIUM, Constants.Q, Constants.HIGH])
+
+   argumentList = parser.parse_args()
+   printParameters(argumentList)
```
Comment (mvdbeek) [inline tools/qr_coder/qr_coder.py]: The print statements are not terribly helpful, could you remove them by default?

File: tools/qr_coder/qr_coder.xml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      <param name="qrVersion"         type="integer" label="An integer from 0 (automatic) to 40 that controls the size of the QR Code" optional="true"  value="0" max="40" min="0" />
```
Comment (mvdbeek) [inline tools/qr_coder/qr_coder.xml]: perhaps add help="Higher numbers produce larger PNG image" ?

**Comment by mvdbeek** [comment]
Nice! Works for me, if you could beautify the few last things I'll merge.

File: tools/qr_coder/qr_coder.xml (hunk 6 above)
Comment (fabiorjvieira) [inline tools/qr_coder/qr_coder.xml]: Done.

File: tools/qr_coder/qr_coder.py (hunk 5 above)
Comment (fabiorjvieira) [inline tools/qr_coder/qr_coder.py]: Done.

File: tools/qr_coder/qr_coder.xml (hunk 4 above)
Comment (fabiorjvieira) [inline tools/qr_coder/qr_coder.xml]: Done. ;)

File: tools/qr_coder/qr_coder.xml (hunk 3 above)
Comment (fabiorjvieira) [inline tools/qr_coder/qr_coder.xml]: Done. ;)

**Comment by mvdbeek** [comment]
Don't forget to upload it in the toolshed of your choice ;)

**Comment by fabiorjvieira** [comment]
If I understood well Christophe wants to upload it to... I do not know where, so I will leave this to him. ;)

**Comment by mvdbeek** [comment]
in that case ping @drosofff

**Pull request merged**